packages/@tailwindcss-upgrade/src/template/codemods/automatic-var-injection.ts

```diff
@@ -0,0 +1,154 @@
+      candidate.modifier?.kind === 'arbitrary' &&
+      !isAutomaticVarInjectionException(designSystem, candidate, candidate.modifier.value)
+    ) {
+      let { value, didChange: modifierDidChange } = injectVar(candidate.modifier.value)
+      candidate.modifier.value = value
+      didChange ||= modifierDidChange
+    }
+
+    // Add `var(…)` to all variants, e.g.:
+    //
+    // `supports-[--test]:flex'` => `supports-[var(--test)]:flex`
+    for (let variant of candidate.variants) {
+      let didChangeVariant = injectVarIntoVariant(designSystem, variant)
+      if (didChangeVariant) {
+        didChange = true
+      }
+    }
+
+    // Add `var(…)` to arbitrary candidates, e.g.:
+    //
+    // `[color:--my-color]` => `[color:var(--my-color)]`
+    if (
+      candidate.kind === 'arbitrary' &&
+      !isAutomaticVarInjectionException(designSystem, candidate, candidate.value)
+    ) {
+      let { value, didChange: valueDidChange } = injectVar(candidate.value)
+      candidate.value = value
+      didChange ||= valueDidChange
+    }
+
+    // Add `var(…)` to arbitrary values for functional candidates, e.g.:
+    //
+    // `bg-[--my-color]` => `bg-[var(--my-color)]`
+    if (
+      candidate.kind === 'functional' &&
+      candidate.value &&
+      candidate.value.kind === 'arbitrary' &&
+      !isAutomaticVarInjectionException(designSystem, candidate, candidate.value.value)
+    ) {
+      let { value, didChange: valueDidChange } = injectVar(candidate.value.value)
+      candidate.value.value = value
+      didChange ||= valueDidChange
+    }
+
+    if (didChange) {
+      return printCandidate(candidate)
+    }
+  }
+  return rawCandidate
+}
+
+function injectVar(value: string): { value: string; didChange: boolean } {
+  let didChange = false
+  if (value.startsWith('--')) {
+    value = `var(${value})`
+    didChange = true
+  } else if (value.startsWith(' --')) {
+    value = value.slice(1)
+    didChange = true
+  }
+  return { value, didChange }
+}
+
+function injectVarIntoVariant(designSystem: DesignSystem, variant: Variant): boolean {
+  let didChange = false
+  if (
+    variant.kind === 'functional' &&
+    variant.value &&
+    variant.value.kind === 'arbitrary' &&
+    !isAutomaticVarInjectionException(
+      designSystem,
+      createEmptyCandidate(variant),
+      variant.value.value,
+    )
+  ) {
+    let { value, didChange: valueDidChange } = injectVar(variant.value.value)
+    variant.value.value = value
+    didChange ||= valueDidChange
+  }
+
+  if (variant.kind === 'compound') {
+    let compoundDidChange = injectVarIntoVariant(designSystem, variant.variant)
+    if (compoundDidChange) {
+      didChange = true
+    }
+  }
+
+  return didChange
+}
+
+function createEmptyCandidate(variant: Variant) {
+  return {
+    kind: 'arbitrary' as const,
+    property: 'color',
+    value: 'red',
+    modifier: null,
+    variants: [variant],
+    important: false,
+    raw: 'candidate',
+  }
+}
```

Comment:
Do we want to make sure that this is always in sync with the `Candidate` type? If so, we can probably do something like:
```suggestion
function createEmptyCandidate(variant: Variant) {
  return {
    kind: 'arbitrary' as const,
    property: 'color',
    value: 'red',
    modifier: null,
    variants: [variant],
    important: false,
    raw: 'candidate',
  } satisfies Candidate
}
```

Or
```suggestion
function createEmptyCandidate(variant: Variant): Candidate {
  return {
    kind: 'arbitrary' as const,
    property: 'color',
    value: 'red',
    modifier: null,
    variants: [variant],
    important: false,
    raw: 'candidate',
  }
}
```